[GTK][WPE] Add WebKitImage utility class to the public API

spenap opened the pull request:
#### f53dba1d3a1d36338bc916c2cecf97205f03b97f
<pre>
[GTK][WPE] Add WebKitImage utility class to the public API
<a href="https://bugs.webkit.org/show_bug.cgi?id=301089">https://bugs.webkit.org/show_bug.cgi?id=301089</a>

Reviewed by Adrian Perez de Castro and Carlos Garcia Campos.

Added a new WebKitImage utility class. This exposes width, height,
stride and image type properties, and allows accessing the native bytes
as well as a GdkTexture for convenience if under PLATFORM(GTK).

This class implements the GIcon and GLoadableIcon interfaces, and has
tests exercising its API as well as the interface implementations.

Tests: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImage.cpp
       Tools/TestWebKitAPI/glib/CMakeLists.txt
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitImage.cpp: Added.
(webkitImageSetProperty):
(webkitImageGetProperty):
(webkit_image_class_init):
(webkitImageNew):
(webkit_image_get_width):
(webkit_image_get_height):
(webkit_image_get_stride):
(webkit_image_as_rgba_bytes):
(webkitImageHash):
(webkitImageEqual):
(webkitImageLoad):
(webkitImageLoadAsync):
(webkitImageLoadFinish):
(webkit_image_gicon_interface_init):
(webkit_image_gloadable_icon_interface_init):
* Source/WebKit/UIProcess/API/glib/WebKitImage.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitImagePrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImage.cpp: Added.
(WebKitImageTest::asyncLoadFinishedCallback):
(WebKitImageTest::WebKitImageTest):
(WebKitImageTest::~WebKitImageTest):
(testWebKitImagePropertiesConstruct):
(testWebKitImageIconInterface):
(testWebKitImageLoadableIconLoadSync):
(testWebKitImageLoadableIconLoadAsync):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:

Canonical link: <a href="https://commits.webkit.org/302720@main">https://commits.webkit.org/302720@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/567c164f698a082fe09de2865889e84149f44398

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129992 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2253 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40852 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137387 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81496 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131863 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2214 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2145 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99016 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/66822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132939 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1644 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116431 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79710 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1560 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34556 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80656 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110080 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35064 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139866 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2047 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1893 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107523 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2092 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112778 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107414 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27343 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1630 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31238 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54869 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2120 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65489 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1935 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1969 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2043 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,333 @@
+ *
+ * Returns: (transfer full): a #WebKitImage
+ */
+WebKitImage* webkit_image_new(int width, int height, guint stride, GBytes* pixels)
```
I would leave this as private. Sure, one can always go through `g_object_new()`, but at the moment we only expect WebKit to create instances of this type. That's surely the case for favicons, and I suspect it will be the same for WebExtensions (cc/ @ItsJamie9494)

ItsJamie9494 replied:
Yes, with WebExtensions it would only be created internally.

spenap replied:
Thanks, I've moved this into the private header

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,333 @@
+    PROP_WIDTH,
+    PROP_HEIGHT,
+    PROP_STRIDE,
+    PROP_PIXELS,
```
I am no completely sure of having a property for the pixel data. My thinking goes as follows:

- To keep it simple (at least for now), we are going to provide the pixel data in a fixed format, which will be BGRA8888 for little-endian targets and ARGB8888 for big-endian ones&mdash;exactly the definition of `GDK_MEMORY_DEFAULT`, so image pixel data can be wrapped as a `GdkTexture` without conversions.
- Getting the `GBytes` with the pixel data _might_ need conversion from the source pixel format to BGRA8888/ARGB8888, which means that the “getter” may not be a cheap operation&mdash;at least the first time, and we can cache the `GBytes` instance, to amortize the conversion when it is needed.
- We may want to use the `webkit_image_get_pixels()` or `_get_bytes()` name in the future, if we ever decide that it would be useful to expose the actual image pixel format with a corresponding `webkit_image_get_format()`&mdash;but this is complexity that IMO we would rather avoid at the moment.
- Therefore, I think calling the method something like `webkit_image_as_rgba_bytes()` instead of `webkit_image_get_pixels()` fits better, without the “get” verb in the name.

aperezdc replied:
All of the above means that we also need a way to customize how to “materialize” the `GBytes` that wraps the pixel data. I think we can have a internal function that WebKit can use to configure a callback. Something like this:

```cpp
using WebKitImageCreateBytesFunction = WTF::Function<RefPtr<GBytes>(WebKitImage*)>;

void webkitImageSetCreateBytesFunction(WebKitImage*, WebKitImageCreateBytesFunction&&);
```

Then the `webkit_image_as_rgba_bytes()` method would either return the `_WebKitImagePrivate::bytes` directly if non-null, or invoke the saved callback and save the resulting value in `::bytes` before returning it.

spenap replied:
Thanks, I've added a `webkit_image_as_rgba_bytes()`  as suggested, and a `WebKitImageCreateBytesFunction` that the constructor receives.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,333 @@
+    int width;
+    int height;
+    unsigned stride;
+    GRefPtr<GBytes> pixels;
```
Let's call it `bytes`, for consistency:


```suggestion
    GRefPtr<GBytes> bytes;
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,333 @@
+    hash = hash * 31 + static_cast<guint>(image->priv->width);
+    hash = hash * 31 + static_cast<guint>(image->priv->height);
+    hash = hash * 31 + static_cast<guint>(image->priv->stride);
+    return hash;
```
Probably better to use `WTF::Hasher` and/or `WTF::computeHash()`, which is type-aware?

```suggestion
    return computeHash(g_bytes_hash(webkit_image_as_rgba_bytes()), image->priv->width, image->priv->height);
```

Also I think that the stride should not matter, two images will display the same as long as their width, height, and pixel contents match&mdash;regardless of how the pixel data is distributed in memory.

spenap replied:
I didn't know about `computeHash` but it feels better, and agree about removing the stride

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,333 @@
+    return image1->priv->width == image2->priv->width
+        && image1->priv->height == image2->priv->height
+        && image1->priv->stride == image2->priv->stride
+        && image1->priv->pixels.get() == image2->priv->pixels.get();
```
Again, I think the stride should not matter here. Also the function is comparing only the pointer of the `GBytes` that wrap the image data; but that can be different for both images and yet the wrapped pixel data be the same.

spenap replied:
Agree -- I've replaced this to use `g_bytes_equal` and removed the stride from the operation

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,333 @@
+
+    auto* image = WEBKIT_IMAGE(icon);
+
+    GInputStream* stream = g_memory_input_stream_new_from_bytes(image->priv->pixels.get());
```
My understanding of `GLoadableIcon` is that it needs to return a stream that has the image data encoded in some image format, like a stream in PNG format, and not raw bytes. Code that uses `g_loadable_icon_load()` won't be able to make sense of the raw pixel data because it does not carry any additional information.

For example `gdk_texture_loadable_icon_load()` saves the texture [contents as PNG format](https://gitlab.gnome.org/GNOME/gtk/-/blob/f76ad9ffec649e1ffa1cd8dc9654ab16ea70f7db/gdk/gdktexture.c#L196) before returning a stream over the produced PNG data.

aperezdc replied:
Also: if we want (or need) to optimize this, in the future we could create internally a `WebKitIconDatabaseImage` subclass of `WebKitImage` with a different implementation that [streams the data from the database blobs](https://sqlite.org/c3ref/blob_open.html) directly (icons in the database are stored in their original format: PNG, JPG, etc.)

spenap replied:
As discussed online, I am exposing an image `type` as a property (which could be PNG or whatever we are told we're storing), and then that type is exposed in the `load` (sync) or in the `load_finish`.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImagePrivate.h:
```diff
@@ -0,0 +1,29 @@
+
+GInputStream* webkit_image_load(GLoadableIcon*, int, char**, GCancellable*, GError**);
+void webkit_image_load_async(GLoadableIcon*, int, GCancellable*, GAsyncReadyCallback, gpointer);
+GInputStream* webkit_image_load_finish(GLoadableIcon*, GAsyncResult*, char**, GError**);
```
This header does not seem to be needed. Virtual function implementations for GObjects are typically `static` to the file that implements a class, in this case `WebKitImage.cpp`.

spenap replied:
Thanks! I've removed these from here, and the private header now only contains the `webkitImageNew` method. I'm using `WK_EXPORT` there as I see we do for methods we want to be visible from the unit test binary, let me know if that's not right

aperezdc reviewed (commented):
I have an important comment to discuss about how to internally allow customizing how to provide access to the pixel data. PTAL.

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,374 @@
+ *
+ * Get the @image type
+ *
+ * Returns: the image type
```
This appears to be `nullable`.

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,374 @@
+
+    auto* image = WEBKIT_IMAGE(icon);
+    if (type)
+        *type = !image->priv->type.isNull() ? g_strdup(image->priv->type.data()) : nullptr;
```
`g_strdup(nullptr)` is safe. Same elsewhere.

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,374 @@
+    GInputStream* stream = g_memory_input_stream_new_from_bytes(webkit_image_as_rgba_bytes(image));
+
+    if (stream && type)
+        *type = !image->priv->type.isNull() ? g_strdup(image->priv->type.data()) : g_strdup("unlucky");
```
`stream` will never be `nullptr`.

spenap replied:
Oh, and the unlucky is a left over from my tests 🙆🏻‍♂️ 🙈

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,374 @@
+        auto* data = static_cast<LoadTaskData*>(taskData);
+        GUniqueOutPtr<GError> error;
+        GRefPtr<GInputStream> stream = webkit_image_load(data->icon, data->size, nullptr, cancellable, &error.outPtr());
+        if (!stream || error)
```
Harmless, but `error` implies `!stream`, so I'd say convention is just `if (error)`

spenap replied:
`webkit_image_load` was not setting `error`. The only possible error path is `webkit_image_as_rgba_bytes` returning null -- because somehow the user-provided function would fail.

I am adding an assert in `webkit_image_as_rgba_bytes`, so that if it fails we know where (I don't think it should be a RELEASE_ASSERT, but I'm open minded).

Then, in `webkit_image_load` I'll do a null check and set error.

TingPing replied:
Ah I assumed the `error` was for future usage.

spenap replied:
Thanks a lot for your review, @TingPing ! Yes, in a way it was for future, but you have a good point: we should say `if (error)` and that doesn't make sense if I never set it, so... 

I hope the updated changes make sense, thanks!

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.h.in:
```diff
@@ -0,0 +1,61 @@
+struct _WebKitImageClass {
+    GObjectClass parent_class;
+};
+#endif
```
Let's remove the definitions inside the `!ENABLE(2022_GLIB_API)` block, because the idea is to define the new `WebKitImage` type only for the 2022+ APIs.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.h.in:
```diff
@@ -0,0 +1,61 @@
+WEBKIT_API const gchar *webkit_image_get_image_type(WebKitImage *image);
+WEBKIT_API GBytes      *webkit_image_as_rgba_bytes (WebKitImage *image);
+#if PLATFORM(GTK)
+WEBKIT_API GdkTexture  *webkit_image_to_gdk_texture(WebKitImage *image);
```
Should we call this `webkit_image_as_gdk_texture()` for consistency with the other method? (cc/ @carlosgcampos)

carlosgcampos replied:
I would not expose this at all.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImagePrivate.h:
```diff
@@ -0,0 +1,28 @@
+
+using WebKitImageCreateBytesFunction = WTF::Function<GRefPtr<GBytes>(WebKitImage*)>;
+
+WK_EXPORT WebKitImage* webkitImageNew(int width, int height, guint stride, const CString& type, WebKitImageCreateBytesFunction&& createBytesFn);
```
I would use `WTF::CompletionHandler`, which is similar to `WTF::Function` but guarantees that it will be only ever invoked once&mdash;which is the case here.

Also, probably not worth it to define the `WebKitImageCreateBytesFunction` type alias, given that it is used only in a few places.


```suggestion
WK_EXPORT WebKitImage* webkitImageNew(int width, int height, guint stride, CompletionHandler<GRefPtr<GBytes>(WebKitImage*)>&& createBytes);
```

Note how the `WTF::` namespace prefix is not needed: most declarations provided by WTF are automatically placed into the global namespace for the sake of convenience, and using the prefix is rarely needed when we need to disambiguate.

spenap replied:
It seems that `CompletionHandler` wants to run on the main thread, but if we do this, we would defeat the purpose of the `load_async`, wouldn't we?

I will check if there's a completion handler equivalent -- which seems to be an std::exchange with null to ensure the single invokation -- that doesn't impose a thread requirement.

I see there's a second argument, `CompletionHandlerCallThread` :+1:

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,381 @@
+
+    GRefPtr<GBytes> bytes;
+};
+
```
It would be nice to add here a paragraph or two explaining what the `WebKitImage` type is. We have such blurbs for most types, see for example [WebKitUserContentManager](https://wpewebkit.org/reference/stable/wpe-webkit-2.0/class.UserContentManager.html) or [WebKitBackForwardList](https://wpewebkit.org/reference/stable/wpe-webkit-2.0/class.BackForwardList.html), and I think it helps understand the API better. Plus, we can indicate in which WebKit version the class was added.


```suggestion
/**
 * WebKitImage:
 *
 * Does beep-meep-zorp and cures all imaginable diseases. It's magic!
 *
 * Since: 2.52
 */
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,381 @@
+    /**
+     * WebKitImage:width:
+     *
+     * The image width
```
...in pixels. It's good to indicate what is the unit of measure.


```suggestion
     * The image width in pixels.
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,381 @@
+    /**
+     * WebKitImage:height:
+     *
+     * The image height
```
```suggestion
     * The image height in pixels.
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,381 @@
+    /**
+     * WebKitImage:stride:
+     *
+     * The image stride
```
It would not hurt to have indicate the units here, and throw in a bit more of information for those using the API that may not know just from the word “stride” what this represents:


```suggestion
     * The image stride, in bytes. This value indicates the amount of
     * memory occupied by each row of pixels in the image. Note that
     * the stride may be larger than the image width multiplied by the
     * amount of bytes used to represent each pixel.
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,381 @@
+ * webkit_image_as_rgba_bytes:
+ * @image: a #WebKitImage
+ *
+ * Get the @image bytes in their native format
```
```suggestion
 * Get the @image pixel data as an array of bytes.
 * 
 * The pixel format for the returned byte buffer is 32-bit per pixel
 * with 8-bit premultiplied alpha, in the preferred byte order for
 * the architecture (typically ABGR8888 on little-endian hosts, and
 * RGBA8888 on big-endian ones).
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,381 @@
+
+    if (!image->priv->bytes)
+        image->priv->bytes = image->priv->createBytesFn(image);
+    ASSERT(image->priv->bytes);
```
I would also assert that we have a valid callable before trying to use to materialize the `GBytes`: 

```suggestion
    if (!image->priv->bytes) {
        ASSERT(image->priv->createBytes);
        image->priv->bytes = image->priv->createBytes(image);
    }
    ASSERT(image->priv->bytes);
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,381 @@
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(icon), 0);
+
+    auto* image = WEBKIT_IMAGE(icon);
+    guint hash = g_bytes_hash(webkit_image_as_rgba_bytes(image));
```
I just realized that the bytes of each row stride that fall “out” of the row pixels should be ignored:

```

   /-------------- stride ---------------\
  /                                       \
 //------- pixels -----------\             \
//                            \             \
+==============================+============+
|  actual                      | potential  |
|      pixel                   |    garbage |              
|         data                 |            |
+==============================+============+
```

We want to use only to hash the actual pixel data. Two images that have the same pixel data should hash to the same value, regardless of the garbage. I'm afraid we will have to traverse the rows and use a `WTF::Hasher` to calculate the hash incrementally.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,381 @@
+
+    return image1->priv->width == image2->priv->width
+        && image1->priv->height == image2->priv->height
+        && g_bytes_equal(webkit_image_as_rgba_bytes(image1), webkit_image_as_rgba_bytes(image2));
```
Same here, `g_bytes_equal()` will also compare the “garbage” after the pixel data in each row stride, and we want to use only the actual pixel data.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,381 @@
+    GInputStream* stream = g_memory_input_stream_new_from_bytes(bytes);
+
+    if (type)
+        *type = g_strdup(image->priv->type.data());
```
This won't work, we need to generate a stream of e.g. PNG data here. The buffer we get from `webkit_image_as_rgba_data()` is raw, does not have any format.

The way to go about it would be to use Skia to wrap the raw data in a `SkData` (similar to `GBytes`), then use `RasterFromData()` to create a `SkImage` that can be encoded to PNG using `SkPngEncoder::Encode()`. At this point we have a new `SkData` with the PNG encoded image, which can be wrapped in `GBytes` and then _that_ is what we can pass to `g_memory_input_stream_new_from_bytes()`.

Of course, then just go ahead and hardcode `*type = "image/png"`.

aperezdc replied:
FWIW, if we want to make something more optimal in the future, for example the `IconDatabase` can subclass `WebKitImage`, and override the, `.load()` vfunc to provide a stream that fetches data directly from the SQLite blob in the database where images are stored encoded 😉

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,381 @@
+    PROP_WIDTH,
+    PROP_HEIGHT,
+    PROP_STRIDE,
+    PROP_TYPE,
```
I would rather *not* have this property at the moment.

Currently in this patch the `type` is used for implementing the `GLoadableIcon` interface that is not really needed because when the `.load()` or `.load_async()` vfuncs are used there is already a mechanism to return the MIME type of the data for stream they return&mdash;there is no need to expose it elsewhere.

aperezdc reviewed (changes requested):
Sorry that it took me a while to get back to this... and also about the churn, because I have again a number of comments that need changes in the patch 😓 

Thanks for your patience with the reviews 🙏🏼

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.h.in:
```diff
@@ -0,0 +1,50 @@
+WEBKIT_API GBytes      *webkit_image_as_rgba_bytes (WebKitImage *image);
+#if PLATFORM(GTK)
+WEBKIT_API GdkTexture  *webkit_image_to_gdk_texture(WebKitImage *image);
+#endif
```
There has to be at least one space before the opening paren (yes, GObject-style header formatting is odd). Also, you forgot to remove the `webkit_image_get_image_type()` function from the header:


```suggestion
WEBKIT_API int          webkit_image_get_width      (WebKitImage *image);
WEBKIT_API int          webkit_image_get_height     (WebKitImage *image);
WEBKIT_API guint        webkit_image_get_stride     (WebKitImage *image);
WEBKIT_API GBytes      *webkit_image_as_rgba_bytes  (WebKitImage *image);
#if PLATFORM(GTK)
WEBKIT_API GdkTexture  *webkit_image_to_gdk_texture (WebKitImage *image);
#endif
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,477 @@
+ * When instantiating this class, stride must be large enough to accommodate the image width times
+ * the bytes-per-pixel representation.
+ * When instantiating this class and providing a CompletionHandler to instantiate the bytes, make sure
+ * this uses #CompletionHandlerCallThread::AnyThread if you first instantiate the class bytes asynchronously.
```
Documentation comments are for users of the public API, but this paragraph is clearly useful for WebKit developers. I would move this below next to the definition of the `webkitImageNew()`, as a regular comment.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,477 @@
+    N_PROPERTIES
+};
+
+enum class ImageFormatBPP : uint8_t {
```
BPP is typically an abbreviation for “**bits** per pixel”, but here is used as “**bytes** per pixel”. This is a bit confusing.

Also, this could be a `static constexpr`:


```suggestion
static constexpr auto RGBA8BytesPerPixel = 4;
```

spenap replied:
I thought about the `static constexpr` but chose the enum in case we added more -- since `GraphicsContextGL` defines so many formats. But I agree -- the constexpr covers what we need

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,477 @@
+    auto* priv = image->priv;
+
+    gsize rowBytes = priv->width * static_cast<uint8_t>(ImageFormatBPP::RGBA8);
+    g_return_if_fail(priv->stride >= rowBytes);
```
This is an internal check, so therefore it should be an assertion&mdash;the `g_return_[val_]if_fail()` macros are intended to be used at the top of public API functions.


```suggestion
    RELEASE_ASSERT(priv->width * RGBA8BytesPerPixel <= priv->stride);
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,477 @@
+WebKitImage* webkitImageNew(int width, int height, guint stride, CompletionHandler<GRefPtr<GBytes>(WebKitImage*)>&& createBytesFn)
+{
+    g_return_val_if_fail(createBytesFn, nullptr);
+
```
We could actually move the assertion from the `.constructed` vfunc here, and remove the `webkitImageConstructed` function altogether. Also, as this is an internal WebKit function, let's replace `g_return_val_if_fail()` with an assertion:


```suggestion
    RELEASE_ASSERT(priv->width * RGBA8BytesPerPixel <= priv->stride);
    RELEASE_ASSERT(createBytesFn);
```

spenap replied:
my reasoning, when I added  the (incorrect) assertion to `.constructed` was that I wanted it to run for both `webkitImageNew` or `g_object_new` users. 

I'm happy to add the `RELEASE_ASSERT(createBytesFn)` here, I think I should add `RELEASE_ASSERT(priv->width * RGBA8BytesPerPixel <= priv->stride);` to the constructed, and keep it. Unless you are quite against the idea of the `constructed` just for that?

aperezdc replied:
I think it's better to remove the `webkitImageConstructed` function.

It would make sense if we would be expecting client code using the public API to create `WebKitImage` instances; but only WebKit will be doing that, and we can remind people to use the `webkitImageNew()` helper function during code review.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,477 @@
+
+    ASSERT(size >= image->priv->stride * (image->priv->height - 1) + rowBytes);
+
+    for (int y = 0; y < image->priv->height; ++y) {
+        auto span = std::span<const uint8_t>(data + y * image->priv->stride, rowBytes);
+        WTF::add(hasher, span);
+    }
```
I think this could use `consumeSpan()` and `span::first()` to avoid the pointer arithmetic altogether:

```suggestion
    auto rowData = span(bytes);
    for (int row = 0; row < image->priv->height; ++row) {
        WTF::add(hasher, rowData.first(rowBytes));
        rowData = consumeSpan(rowData, image->priv->stride);
    }
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,477 @@
+    auto* image1 = WEBKIT_IMAGE(icon1);
+    auto* image2 = WEBKIT_IMAGE(icon2);
+
+    if (image1->priv->width != image2->priv->width
+        || image1->priv->height != image2->priv->height)
```
Surprising that the style checker does not complain about this suggesting to put the whole condition in a single line 🤷🏼

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,477 @@
+    ASSERT(size1 >= image1->priv->stride * (image1->priv->height - 1) + rowBytes);
+    ASSERT(size2 >= image2->priv->stride * (image1->priv->height - 1) + rowBytes);
+
+    for (int y = 0; y < image1->priv->height; ++y) {
+        auto span1 = std::span<const uint8_t>(data1 + y * image1->priv->stride, rowBytes);
+        auto span2 = std::span<const uint8_t>(data2 + y * image2->priv->stride, rowBytes);
+
+        if (!equalSpans(span1, span2))
+            return false;
+    }
+
```
I think this could be written using `consumeSpan()` and `span::first()`, in a similar way as suggested for the implementation of the `.hash` vfunc.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,477 @@
+
+static GInputStream* webkit_image_load(GLoadableIcon*, int, char**, GCancellable*, GError**);
+static void webkit_image_load_async(GLoadableIcon*, int, GCancellable*, GAsyncReadyCallback, gpointer);
+static GInputStream* webkit_image_load_finish(GLoadableIcon*, GAsyncResult*, char**, GError**);
```
We use `dromedaryCase` for internal functions:


```suggestion
static guint webkitImageHash(GIcon*);
static gboolean webkitImageEqual(GIcon*, GIcon*);

static GInputStream* webkitImageLoad(GLoadableIcon*, int, char**, GCancellable*, GError**);
static void webkitImageLoadAsync(GLoadableIcon*, int, GCancellable*, GAsyncReadyCallback, gpointer);
static GInputStream* webkitImageLoadFinish(GLoadableIcon*, GAsyncResult*, char**, GError**);
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,477 @@
+    sk_sp<SkImage> skImage = SkImages::RasterFromData(info, skData, image->priv->stride);
+
+    if (!skImage) {
+        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to create SkImage");
```
I suppose we do not want to let client code using the API know that we are using Skia here.


```suggestion
        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to encode PNG");
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,477 @@
+        return nullptr;
+    }
+
+    pngData->ref();
+    GRefPtr<GBytes> pngBytes = adoptGRef(g_bytes_new_with_free_func(
+        pngData->data(),
+        pngData->size(),
+        [](gpointer userData) {
+            static_cast<SkData*>(userData)->unref();
+        },
+        pngData.get()
+    ));
```
We can skip the reference count churn by telling the `sk_sp` to “release” the ownwership of the object and transfer it to the GBytes;

```suggestion
    GRefPtr<GBytes> pngBytes = adoptGRef(g_bytes_new_with_free_func(
        pngData->data(),
        pngData->size(),
        [](gpointer userData) {
            static_cast<SkData*>(userData)->unref();
        },
        pngData.release()
    ));
```

spenap replied:
I had previously tried this, but without a debug build, so I couldn't tell exactly where it was crashing. With a debug build now, the problem with `pngData.release()` is that it sets its pointed data to null, so `pngData->data()` or `pngData->size()` will do a null pointer dereference.

The extra ref at the top is matched with the unref in the free_func, and then the object is normally disposed as we exit `load`

aperezdc reviewed (commented):
Thanks @spenap! This is looking great and almost ready to go. I left some comments but they are only small nits at this point.

As this adds new API, it would be good to have a second review, ideally by people who typically have worked with the public API before, like @carlosgcampos, @mcatanzaro and/or @TingPing.

If @ItsJamie9494 could take a look to make sure that we do not miss anything they may need for the WebExtensions work, that would be a very nice bonus.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+#else
+#include <WebCore/NotImplemented.h>
+#endif
+#include <span>
```
This line shouldn't be needed, because it will be transitively included by `GSpanExtras.h`.


```suggestion
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+ *
+ * Since: 2.52
+ */
+GBytes* webkit_image_as_rgba_bytes(WebKitImage* image)
```
I'm sorry if this was discussed before, but if we plan to support more formats why don't we have a single getter get_data that receives an enum that for now only contains RGBA? If we eventually add support for other formats, the only new api is an enum value

aperezdc replied:
The rationale here is trying to provide the smallest possible interface that is convenient to use. In particular, it seemed worth it not having to push to client code the responsibility of having to handle multiple pixel formats.

What seemed reasonable was to provide pixel data in a fixed format that would be suitable for the use cases we have in mind (page icons, WebExtension icons, web view screenshots) and most of the time would not need format conversions: 32bpp RGBA, in “native” order (BGRA for little-endian, ARGB for big-endian&mdash;the same as for `GDK_MEMORY_DEFAULT`). The reason for naming this with `_as_rgba_bytes` is to encode that notion, and the idea that conversion _might_ happen (otherwise we would use “get” in the name: just a getter). 

We already have the precedent of specifying a fixed pixel format for [wpe_view_set_cursor_from_bytes()](https://github.com/WebKit/WebKit/blob/4ef19b5fa4c093b11b221c8d8df3b5a3d9dd398d/Source/WebKit/WPEPlatform/wpe/WPEView.cpp#L847).

If we ever want to extend the interface, what would make sense is to allow direct access to the data without conversions, and for that we would add a plain getter for the pixel format, and another getter for the buffer data. Something like these:

```c
typedef enum { /* ... */ } WebKitImageFormat;
WebKitImageFormat  webkit_image_get_format (WebKitImage *image);
GBytes            *webkit_image_get_bytes  (WebKitImage *image);
```

Adding these later will not conflict with having an `_as_rgba_bytes()` operation.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+
+    auto* bytes = webkit_image_as_rgba_bytes(image);
+
+    return gdk_memory_texture_new(image->priv->width, image->priv->height, GDK_MEMORY_DEFAULT, bytes, image->priv->stride);
```
I'm not sure it's worth exposing this, since the caller can just use gdk_memory_texture_new

aperezdc replied:
Covenience, but we can leave it out.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+    iface->load_finish = webkitImageLoadFinish;
+}
+
+guint webkitImageHash(GIcon* icon)
```
```suggestion
static guint webkitImageHash(GIcon* icon)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+#endif
+
+static guint webkitImageHash(GIcon*);
+static gboolean webkitImageEqual(GIcon*, GIcon*);
```
Is it possible to move the implementation here instead of declaring prototypes?

spenap replied:
I've done that, thanks

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+ *
+ * At the moment, this assumes native images are in RGBA format, with 4 bytes per pixel.
+ * When instantiating this class, stride must be large enough to accommodate the image width times
+ * the bytes-per-pixel representation.
```
```suggestion
 * Represents an image as a buffer containing pixel data.
 * 
 * Image objects are always created by WebKit, and considered immutable:
 * a copy of the image data needs to be made before modifying the image.
 * Pixel data can be obtained with [id@webkit_image_as_rgba_bytes].
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+ * webkit_image_get_width:
+ * @image: a #WebKitImage
+ *
+ * Get the @image width in pixels
```
```suggestion
 * Get the @image width in pixels.
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+ * webkit_image_get_height:
+ * @image: a #WebKitImage
+ *
+ * Get the @image height in pixels
```
```suggestion
 * Get the @image height in pixels.
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+/**
+ * webkit_image_get_width:
```
Conversely, let's annotate this to indicate which property this getter returns:


```suggestion
 * webkit_image_get_width: (get-property width):
```

(The same goes for the getter functions of the `height` and `stride` properties.)

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+    objectClass->get_property = webkitImageGetProperty;
+
+    /**
+     * WebKitImage:width:
```
We can indicate the getter of the property, to get better introspection bindings:


```suggestion
     * WebKitImage:width: (getter get_width) (attributes org.gtk.Property.get=webkit_image_get_width):
```

(The same goes for the `height` and `stride` properties.)

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+ * webkit_image_get_stride:
+ * @image: a #WebKitImage
+ *
+ * Get the @image stride
```
```suggestion
 * Get the @image stride.
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+
+    ASSERT(dataSpan.size() >= image->priv->stride * (image->priv->height - 1) + rowBytes);
+
+    for (int y = 0; y < image->priv->height; ++y) {
+        auto row = consumeSpan(dataSpan, image->priv->stride);
+        WTF::add(hasher, row.first(rowBytes));
```
Why don't we just use g_bytes_hash?

spenap replied:
An earlier version used `g_bytes_hash`. The problem is that it will consider the entirety of the bytes, including any possible padding between the end of the row data and the stride.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+        WTF::add(hasher, row.first(rowBytes));
+    }
+
+    WTF::add(hasher, image->priv->width, image->priv->height);
```
Do we really need to also use the size? the bytes hash will be different if the size changes

spenap replied:
I don't have a strong preference either way

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+    ASSERT(dataSpan1.size() >= image1->priv->stride * (image1->priv->height - 1) + rowBytes);
+    ASSERT(dataSpan2.size() >= image2->priv->stride * (image1->priv->height - 1) + rowBytes);
+
+    for (int y = 0; y < image1->priv->height; ++y) {
+        auto row1 = consumeSpan(dataSpan1, image1->priv->stride);
+        auto row2 = consumeSpan(dataSpan2, image2->priv->stride);
+
+        if (!equalSpans(row1.first(rowBytes), row2.first(rowBytes)))
+            return false;
+    }
```
Can we use g_bytes_equal?

spenap replied:
Same as with `g_bytes_hash`, `g_bytes_equal` would consider the padding

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+static void webkit_image_gicon_interface_init(GIconIface* iface)
+{
+    iface->hash = webkitImageHash;
+    iface->equal = webkitImageEqual;
```
Don't we want to implement serialize too?

spenap replied:
We left out serialize and to_tokens / from_tokens. If there's a need we could add it later.

aperezdc reviewed (approved):
One last round of (ever smaller) nits. Patch LGTM overall.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+    return true;
+}
+
+GInputStream* webkitImageLoad(GLoadableIcon* icon, int size, char** type, GCancellable* cancellable, GError** error)
```
```suggestion
static GInputStream* webkitImageLoad(GLoadableIcon* icon, int size, char** type, GCancellable* cancellable, GError** error)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+};
+WEBKIT_DEFINE_ASYNC_DATA_STRUCT(LoadTaskData)
+
+void webkitImageLoadAsync(GLoadableIcon* icon, int size, GCancellable* cancellable, GAsyncReadyCallback callback, gpointer userData)
```
```suggestion
static void webkitImageLoadAsync(GLoadableIcon* icon, int size, GCancellable* cancellable, GAsyncReadyCallback callback, gpointer userData)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,476 @@
+    g_task_run_in_thread(task.get(), taskHandler);
+}
+
+GInputStream* webkitImageLoadFinish(GLoadableIcon* icon, GAsyncResult* res, char** type, GError** error)
```
```suggestion
static GInputStream* webkitImageLoadFinish(GLoadableIcon* icon, GAsyncResult* res, char** type, GError** error)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImagePrivate.h:
```diff
@@ -0,0 +1,27 @@
+#include <wtf/CompletionHandler.h>
+#include <wtf/glib/GRefPtr.h>
+
+WK_EXPORT WebKitImage* webkitImageNew(int width, int height, guint stride, CompletionHandler<GRefPtr<GBytes>(WebKitImage*)>&& createBytesFn);
```
It's difficult for me to understand why we need the completion handler, without seeing how this is used from webkit.

aperezdc replied:
If you want to see an example of how this used, I have made a [quick patch](https://github.com/WebKit/WebKit/pull/53480/commits/e782ba9994664957d903c6b1e3797dc182c398cf) implementing `webkit_web_view_get_snapshot()` on top of this. In that patch the usage is straightforward because we already have the data in the format we want, but the idea is that code creating a `WebKitImage` may need to do extra work to fetch or convert the pixel data, and in that case this approach allows to create the `WebKitImage` right away, and use the completion handler to “materialize” the data lazily, when the user of the image first needs it.

carlosgcampos replied:
That patch doesn't look simple to me, rather the opposite, the completion handler makes it complex for no reason. Since this is internal API, I think it's better to keep it simple and if we ever need to create an image in a different way we can change it. For now it can just receive a GBytes instead of a completion handler.

spenap replied:
I'm removing the completion handler for now, passing bytes directly

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImage.cpp:
```diff
@@ -0,0 +1,243 @@
+#include "config.h"
+
+#include "TestMain.h"
+#include <WebKitImagePrivate.h>
```
Can we test this with images created from WebKit instead of using private api from unit tests?

spenap replied:
We expose `webkitImageNew` in the private header, which is right now the clean way to satisfy the completion handler and provide the bytes.

We haven't updated anything yet to use WebKitImage, but I guess that's not what you mean. Could you give an example of what you have in mind?

carlosgcampos replied:
It's just that I don't understand the api, and since it's not used yet by webkit itself, it's even more difficult for me to understand. When I saw a completion handler in the constructor I thought it was a destroy notify to release any data attached to the image, but it's a function to provide the data? Why do we need a function for that? and why do we use completion handler?

spenap replied:
It allow us to be _lazy_ about the image data, avoiding to copy it unless it is used. That could work with any callback, but specifically using a completion handler guarantees that this is only invoked once

carlosgcampos replied:
Is there any wip patch where this is used? Because I still don't understand it.

aperezdc replied:
Once we make the new page icons API use `WebKitImage`, or enabling the page screenshot API in WPE, we can add tests that exercise images created by WebKit.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -0,0 +1,438 @@
+    g_object_class_install_properties(objectClass, N_PROPERTIES, sObjProperties.data());
+}
+
+WebKitImage* webkitImageNew(int width, int height, guint stride, GRefPtr<GBytes> bytes)
```
```suggestion
WebKitImage* webkitImageNew(int width, int height, guint stride, GRefPtr<GBytes>&& bytes)
```

spenap replied:
This suggestion **forces** users of the API to transfer ownership of the bytes into the image.

The current signature is indicating that both approaches (sharing ownership and transferring ownership) are valid.

Are there any cases where we might want to share ownership or will we always transfer it? @aperezdc , do you have an opinion?

aperezdc replied:
I think it's enough for now to receive a `GRefPtr<GBytes>&&` paramer, as suggested by @carlosgcampos.

aperezdc replied:
By the way, one will still be able do this to share the `GBytes`:

```
GRefPtr<GBytes> otherRef = getBytesFromSomewhereElse();
return webkitImageNew(WTFMove(otherRef));
```

Remember: what gets moved is the smart pointer itself, the wrapper `GBytes` can always get the reference count increased by creating a copy of the _smart pointer_ wrapping it 😉

spenap replied:
Yes, but if you're doing that, you're working around the API, not _with_ the API :-D. In any way, we don't have to address this if it is not a problem.

spenap committed:
[GTK][WPE] Add WebKitImage utility class to the public API
https://bugs.webkit.org/show_bug.cgi?id=301089

Reviewed by Adrian Perez de Castro and Carlos Garcia Campos.

Added a new WebKitImage utility class. This exposes width, height,
stride and image type properties, and allows accessing the native bytes
as well as a GdkTexture for convenience if under PLATFORM(GTK).

This class implements the GIcon and GLoadableIcon interfaces, and has
tests exercising its API as well as the interface implementations.

Tests: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImage.cpp
       Tools/TestWebKitAPI/glib/CMakeLists.txt
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitImage.cpp: Added.
(webkitImageSetProperty):
(webkitImageGetProperty):
(webkit_image_class_init):
(webkitImageNew):
(webkit_image_get_width):
(webkit_image_get_height):
(webkit_image_get_stride):
(webkit_image_as_rgba_bytes):
(webkitImageHash):
(webkitImageEqual):
(webkitImageLoad):
(webkitImageLoadAsync):
(webkitImageLoadFinish):
(webkit_image_gicon_interface_init):
(webkit_image_gloadable_icon_interface_init):
* Source/WebKit/UIProcess/API/glib/WebKitImage.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitImagePrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImage.cpp: Added.
(WebKitImageTest::asyncLoadFinishedCallback):
(WebKitImageTest::WebKitImageTest):
(WebKitImageTest::~WebKitImageTest):
(testWebKitImagePropertiesConstruct):
(testWebKitImageIconInterface):
(testWebKitImageLoadableIconLoadSync):
(testWebKitImageLoadableIconLoadAsync):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:

Canonical link: https://commits.webkit.org/302720@main
